src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel_Navigation.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.Navigation;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    internal partial class DocumentOutlineViewModel
+    {
+        /// <summary>
+        /// A Queue of text spans to navigate to.
+        /// </summary>
+        private readonly AsyncBatchingWorkQueue<TextSpan> _navigationQueue;
+
+        internal void EnqueueNavigation(TextSpan textSpan)
+            => _navigationQueue.AddWork(textSpan, cancelExistingWork: true);
+
+        /// <summary>
+        /// Event handler that is called once we have finished navigating to a span.
+        /// </summary>
+        public event EventHandler? NavigationCompleted;
+
+        private async ValueTask NavigateToTextSpanAsync(ImmutableSegmentedList<TextSpan> textSpans, CancellationToken token)
+        {
+            var textSpan = textSpans.Last();
+            var textView = await _visualStudioCodeWindowInfoService.GetLastActiveIWpfTextViewAsync(token).ConfigureAwait(false);
+            var snapShot = textView.TextSnapshot;
+            var document = snapShot.GetOpenDocumentInCurrentContextWithChanges();
```

Comment:
this is all horribly unsafe :)  if the active view changed, and you're now processing the navigation request, you'll have a *different* document that doesn't correspond to the textspan you're navigating to.

the documentid and text span should be args here, nto just the text span.